src/cpu_feature/arm_linux/arm_linux.rs

```diff
@@ -0,0 +1,514 @@
+    let hwcap_features: ArmHwcapFeatures<AuxvUnsignedLongNative>
+        = ArmHwcapFeatures::new();
+    if let Ok(c) = parse_cpuinfo() {
+        // if we can't load procfs auxv, just let it be empty
+        let procfs_auxv = match
+                auxv::search_procfs_auxv::<AuxvUnsignedLongNative, NativeEndian>
+                    (&Path::from("/proc/self/auxv"),
+                     &[auxv_types.AT_HWCAP, auxv_types.AT_HWCAP2]) {
+            Ok(auxv) => auxv,
+            Err(_) => AuxVals::<AuxvUnsignedLongNative>::new()
+        };
+
+        let getauxval = NativeGetauxvalProvider{};
+        let armcap =
+            armcap_bits::<NativeGetauxvalProvider>(&c, &procfs_auxv, auxv_types,
+                                                   hwcap_features, getauxval);
+        unsafe {
+            GFp_armcap_P |= armcap;
+        }
+    }
+}
+
+/// returns a u32 with bits set for use in GFp_armcap_P
+#[cfg(any(test, all(any(target_arch = "arm", target_arch = "aarch64"),
+    target_os="linux")))]
+fn armcap_bits<G: GetauxvalProvider> (cpuinfo: &CpuInfo,
+                                      procfs_auxv: &AuxVals<AuxvUnsignedLongNative>,
+                                      auxval_types: AuxvTypes<AuxvUnsignedLongNative>,
+                                      hwcap_features: ArmHwcapFeatures<AuxvUnsignedLongNative>,
+                                      getauxval_provider: G)
+        -> u32 {
+    let mut hwcap = AuxvUnsignedLongNative::from(0_u32);
+
+    // |getauxval| is not available on Android until API level 20. If it is
+    // unavailable, read from /proc/self/auxv as a fallback. This is unreadable
+    // on some versions of Android, so further fall back to /proc/cpuinfo.
+    // See
+    // https://android.googlesource.com/platform/ndk/+/882ac8f3392858991a0e1af33b4b7387ec856bd2
+    // and b/13679666 (Google-internal) for details. */
+
+    if let Some(v) = getauxval_provider.getauxval(auxval_types.AT_HWCAP) {
+        hwcap = v;
+    } else if let Some(v) = procfs_auxv.get(&auxval_types.AT_HWCAP) {
+        hwcap = *v;
+    } else if let Some(v) = hwcap_from_cpuinfo(&cpuinfo, hwcap_features) {
+        hwcap = v;
+    }
+
+    // Clear NEON support if known broken
+    if cpu_has_broken_neon(&cpuinfo) {
+        hwcap &= !hwcap_features.ARM_HWCAP_NEON;
+    }
+
+    // Matching OpenSSL, only report other features if NEON is present
+    let mut armcap: u32 = 0;
+    if hwcap & hwcap_features.ARM_HWCAP_NEON > AuxvUnsignedLongNative::from(0_u32) {
+        armcap |= ARMV7_NEON;
+
+        // Some ARMv8 Android devices don't expose AT_HWCAP2. Fall back to
+        // /proc/cpuinfo. See https://crbug.com/596156
+
+        let mut hwcap2 = AuxvUnsignedLongNative::from(0_u32);
+        if let Some(v) = getauxval_provider.getauxval(auxval_types.AT_HWCAP2) {
+            hwcap2 = v;
+        } else if let Some(v) = procfs_auxv.get(&auxval_types.AT_HWCAP2) {
```

Comment:
The original C did not check procfs auxv for `HWCAP2` if `getauxval` was unavailable. I can't imagine why we wouldn't check that, so I put it in, but maybe you can think of why that would be undesirable?